File: ios/StudyApp/Data/TimetableOverdueCalculator.swift
import CoreData
import Foundation

/// Computes the number of timetable lessons whose review is overdue.
///
/// Iterates every active term, walks each day in range, and cross-references
/// entries and review records. Extracted out of `PersistenceController` so it
/// can run on a background context and avoid blocking the UI while the
/// potentially hundreds of dates × entries are scanned.
enum TimetableOverdueCalculator {

    /// Counts lessons whose scheduled end is more than 48 hours before
    /// `reference` and for which no `isReviewed` or `isExcluded` review record
    /// exists.
    static func overdueCount(
        reference: Date,
        in context: NSManagedObjectContext
    ) throws -> Int {
        let terms = try CoreDataQuery.fetch("TimetableTermRecord", in: context)
            .map(PersistenceMappers.timetableTerm)
            .filter { $0.deletedAt == nil && $0.isActive }
        let periods = try CoreDataQuery.fetch("TimetablePeriodRecord", in: context)
            .map(PersistenceMappers.timetablePeriod)
            .filter { $0.deletedAt == nil && $0.isActive }
        let entries = try CoreDataQuery.fetch("TimetableEntryRecord", in: context)
            .map(PersistenceMappers.timetableEntry)
            .filter { $0.deletedAt == nil }
        let reviews = try CoreDataQuery.fetch("TimetableReviewRecord", in: context)
            .map(PersistenceMappers.timetableReviewRecord)
            .filter { $0.deletedAt == nil }

        let periodMap = Dictionary(uniqueKeysWithValues: periods.map { ($0.id, $0) })
        let reviewMap = reviews.reduce(into: [String: TimetableReviewRecord]()) { result, review in
            let key = "\(review.termId)-\(review.entryId)-\(review.periodId)-\(review.occurrenceDate)"
            if let existing = result[key], existing.updatedAt > review.updatedAt {
                return
            }
            result[key] = review
        }
        let calendar = Calendar.current
        var overdue = 0

        for term in terms {
            var date = term.startDateValue
            let lastDate = min(term.endDateValue, reference.startOfDay)
            while date <= lastDate {
                let occurrenceDate = date.epochDay
                let weekday = StudyWeekday.from(calendarWeekday: calendar.component(.weekday, from: date))
                for entry in entries where (entry.termId == term.id || entry.termId == nil) && entry.dayOfWeek == weekday {
                    if let validFromDate = entry.validFromDate, occurrenceDate < validFromDate { continue }
                    if let validToDate = entry.validToDate, occurrenceDate > validToDate { continue }
                    guard let period = periodMap[entry.periodId] else { continue }
                    let key = "\(term.id)-\(entry.id)-\(period.id)-\(occurrenceDate)"
                    if let review = reviewMap[key], review.isReviewed || review.isExcluded {
                        continue
                    }
                    let endDate = calendar.date(
                        bySettingHour: period.endMinute / 60,
                        minute: period.endMinute % 60,
                        second: 0,
                        of: date
                    ) ?? date
                    if reference >= endDate.addingTimeInterval(48 * 60 * 60) {
                        overdue += 1
                    }
                }
                guard let next = calendar.date(byAdding: .day, value: 1, to: date) else { break }
                date = next
            }
        }

        return overdue
    }
}

File: android/app/src/main/java/com/studyapp/domain/usecase/TimetableOverdueCalculator.kt
package com.studyapp.domain.usecase

import com.studyapp.domain.model.StudyWeekday
import com.studyapp.domain.model.TimetableEntry
import com.studyapp.domain.model.TimetablePeriod
import com.studyapp.domain.model.TimetableReviewRecord
import com.studyapp.domain.model.TimetableTerm
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.LocalTime

object TimetableOverdueCalculator {
    private const val OVERDUE_HOURS = 48L

    fun overdueCount(
        reference: LocalDateTime,
        terms: List<TimetableTerm>,
        periods: List<TimetablePeriod>,
        entries: List<TimetableEntry>,
        reviews: List<TimetableReviewRecord>
    ): Int {
        val activePeriods = periods
            .filter { it.isActive && it.deletedAt == null && it.startMinute < it.endMinute }
            .sortedWith(compareBy<TimetablePeriod> { it.sortOrder }.thenBy { it.startMinute })
        if (activePeriods.isEmpty()) return 0

        val periodMap = activePeriods.associateBy { it.id }
        val reviewMap = reviews
            .filter { it.deletedAt == null }
            .groupBy { "${it.termId}-${it.entryId}-${it.periodId}-${it.occurrenceDate}" }
            .mapValues { (_, records) -> records.maxByOrNull { it.updatedAt } }
        val referenceDate = reference.toLocalDate()
        var overdue = 0

        for (term in terms.filter { it.deletedAt == null && it.isActive }) {
            var date = term.startDateValue
            val lastDate = minOf(term.endDateValue, referenceDate)
            while (!date.isAfter(lastDate)) {
                val occurrenceDate = date.toEpochDay()
                val weekday = StudyWeekday.fromDayOfWeek(date.dayOfWeek)
                for (entry in entries) {
                    if (entry.deletedAt != null) continue
                    if (entry.termId != term.id && entry.termId != null) continue
                    if (entry.dayOfWeek != weekday) continue
                    if (entry.validFromDate != null && occurrenceDate < entry.validFromDate) continue
                    if (entry.validToDate != null && occurrenceDate > entry.validToDate) continue
                    val period = periodMap[entry.periodId] ?: continue
                    val key = "${term.id}-${entry.id}-${period.id}-$occurrenceDate"
                    val review = reviewMap[key]
                    if (review?.isReviewed == true || review?.isExcluded == true) continue
                    val endDateTime = LocalDateTime.of(
                        date,
                        LocalTime.of(period.endMinute / 60, period.endMinute % 60)
                    )
                    if (!reference.isBefore(endDateTime.plusHours(OVERDUE_HOURS))) {
                        overdue += 1
                    }
                }
                date = date.plusDays(1)
            }
        }

        return overdue
    }
}
